### ISSUE TITLE
problem on /usr/local/opnsense/scripts/openvpn/ovpn_event.py 

### ISSUE THREAD

**Issue by amastran** [opnsense/core#6976]
good morning,
I experience a problem with openvpn OTP:
the problem is when this file is executed:
/usr/local/opnsense/scripts/openvpn/ovpn_event.py

specifically this routine:

elif params.script_type == 'learn-address':
        if os.fork() == 0:
            sys.exit(subprocess.run(
                ['/usr/local/opnsense/scripts/filter/update_tables.py', '--types', 'authgroup']
            ).returncode)

That procedure is called from this file
 /usr/local/opnsense/scripts/filter/update_tables.py

when executed the cpu is saturated, with the python process running at 100% and packet loss is experienced making the vpn unusable.

For now I solved it by changing the name "learn-address" to something that doesn't exist.

I would like to know how I can solve the problem

Thank you

**Comment by AdSchellevis** [comment]
usually `/usr/local/opnsense/scripts/filter/update_tables.py --types authgroup` shouldn't take very long and 'learn-address' should only be invoked when addresses are being added to the internal routing table (https://openvpn.net/community-resources/reference-manual-for-openvpn-2-6/).

The function itself is responsible for populating these aliases https://docs.opnsense.org/manual/aliases.html#openvpn-group

You can always run the script in a console and see if it does take long on your end for some unforseen reason.

**Comment by amastran** [comment]
The problem is that when this script runs, the CPU goes to 100% on all cores, making the VPN unusable due to packet loss. I use openvpn with 300 users. is it possible to somehow delete the __openvpn_network alias considering that pf is turned off?

**Comment by AdSchellevis** [comment]
it's not that alias, but if the functionality isn't used, we might consider being able to toggle this. it is a bit odd though, pythin itself can only use a single core (which can't cause any disruption on a multicore machine)

**Comment by amastran** [comment]
with old versions of opnsense this didn't happen, I experimented with 200 users 4% packet loss. It's not an internet connection problem. I was thinking of increasing the number of udp datagrams in sysctl? it's a good idea?

**Comment by AdSchellevis** [comment]
well, openvpn is single core as well, so with a lot of users, it sometimes makes sense to use different servers to split the load. If you run `time /usr/local/opnsense/scripts/filter/update_tables.py --types authgroup` , what does it return?

**Comment by amastran** [comment]
root@sii-opnsense04:~ # time /usr/local/opnsense/scripts/filter/update_tables.py --types authgroup
{"status": "ok"}
0.692u 0.063s 0:01.41 53.1%     437+172k 0+1io 0pf+0w 

as a workaround I changed the code, putting "elif params.script_type == 'XX-learn-address':
        if os.fork() == 0:
            sys.exit(subprocess.run(
                ['/usr/local/opnsense/scripts/filter/update_tables.py', '--types', 'authgroup']
            ).returncode)
in this way the problem does not occur. naturally the check you asked me to do was done with the original file

**Comment by amastran** [comment]
the problem seems to be " elif params.script_type == 'learn-address':"

**Comment by amastran** [comment]
Hi AdSchellevis, do you have anything else to suggest? Do you have any best practices to suggest for minimizing packet loss? I'll summarize by saying that I've been using the openvpn OTP implementation with active directory for about 2 years with two redundant servers and that the old versions of opnsense didn't have these problems. I currently use: 1 dec480, 2 virtual machines to distribute the load, with a percentage of 130 users on the dec and 40 on the VMs, with a total of 3 WAN. with the modification of "elif params.script_type == 'learn-address':" of the file "/usr/local/opnsense/scripts/openvpn/ovpn_event.py" the situation improves drastically even if it still presents packet loss problems which with I didn't have the old versions. I got a DEC480 hoping it would improve performance.

 openvpn --version
OpenVPN 2.6.6 amd64-portbld-freebsd13.2 [SSL (OpenSSL)] [LZO] [LZ4] [PKCS11] [MH/RECVDA] [AEAD]
library versions: OpenSSL 1.1.1w 11 Sep 2023, LZO 2.10

**Comment by AdSchellevis** [comment]
not really, no, because I don't know the cause of your issues. my assumption is that your problem is similar when the event isn't there, which might mean you're at the maximum a single openvpn server can handle (for being single threaded) or there's some other configuration issue. Unfortunately this lies a bit out of community support scope.
